rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,416 @@
+    T *data() const {
+        return static_cast<T *>(data_);
+    }
+
+    size_t size() const { return size_; }
+
+protected:
+    void *data_;
+    size_t size_;
+    const dnnl::engine eng_;
+    const graph::allocator_t *alc_;
+
+private:
+    malloc_func_t malloc_func_;
+    free_func_t free_func_;
+};
+```
+
+The constant tensor cache class may looks like the following.
+
+```cpp
+class unified_constant_tensor_cache_t {
+public:
+    using key_t = size_t;
+    using cached_t = std::shared_ptr<constant_buffer_t>;
+    using value_t = std::shared_future<cached_t>;
+    using promise_t = std::promise<cached_t>;
+
+    unified_constant_tensor_cache_t();
+    ~unified_constant_tensor_cache_t();
+
+    status_t set_capacity(size_t capacity);
+    size_t get_capacity();
+
+    size_t get_size() const;
+
+    /// @backend_id:
+    /// @backend_specific_key:
+    /// @size: The size the constant tensor
+    /// @value: A promised future constant tensor.
+    /// @return A valid future of cache hit, and an invalid one if cache miss.
+    value_t get_or_add(size_t backend_id, size_t backend_specific_key,
+            sze_t size, const value_t &value);
+
+    void remove_if_exist(size_t backend_id, size_t backend_specific_key);
+};
+
+template<typename engine_kind>
+unified_constant_tensor_cache_t &get_unified_constant_tensor_cache_t();
+```
+
+The `set_capacity` and `get_capacity`are the
+implementations of the frontend capacity control APIs.
+
+The `set_capacity` must not be used by backends. The `get_capacity` can be used
+by backends to query information. Besides, the The following interfaces are also
+provided to backends as a part of backend API:
+
+- `get_or_add`: this API is used to query a constant tensor from the cache. When
+  using this API, backend should pass in an empty `value` argument. If cache
+  hit, the returned `value_t` will be valid, and the passed empty `value` will
+  be ignored. If cache miss, the returned `value_t` will be invalid and the
+  behavior for empty `value` is different depending on whether the cache is full
+  or not. We will elaborate the cache miss situation later.
+- `remove_if_exist`: this API is used to remove a tensor from the cache. If the
+  removed tensor is currently being used, backend should hold it and make sure
+  it's alive.
+- `get_size`: this API is used to query the total size of all cached tensors.
+
+The `get_or_add` API behaviors when cache miss:
+
+If cache is not full, the empty `value` will be added to cache as a promise, and
+then the backend compute the constant tensor and set it to the empty `value` to
+fulfill the promise. In concurrent execution, if multiple threads query cache
+with same key, only one thread will cache miss and do the computation, other
+threads will cache hit and wait the promise to be fulfilled. This behavior is
+exactly same as the existing constant cache behaviors in DNNL backend.
+
+If cache is full, the behavior may be different for different policies:
```

Comment:
No need to describe the policies? As we don't provide them in API and no guarantee for implementation?

Reply:
removed